service/project.go

```diff
@@ -338,31 +338,56 @@ func (s *ProjectService) CancelInvitation(ctx context.Context, projectID, invita
 }
 
 func (s *ProjectService) AcceptInvitation(ctx context.Context, tkn cryptox.Token) error {
-	invitation, err := s.getPendingInvitationByToken(ctx, tkn)
+	// When accepting an invitation, two possible scenarios can happen:
+	// 1. User does not exist yet, in which case invitation is only accepted
+	//    When a user registers later, a project member will be created (PostgreSQL function handle_new_user() is triggered upon user creation)
+	//    Must be done via Postgres function because user registration is not controlled by this API
+	// 2. User already exists, in which case a project member is created and invitation is deleted
+	//
+	// To keep more logic in service, we first fetch the invitation to get invitation's email
+	// Then we check if user with given email exists
+	// First two steps could be done in one query, but it would require mixing entities from two different repositories
+	// Based on user existence we either accept the invitation or create a project member and delete the invitation
+	//
+	// There is one possible raise condition:
+	// 1. We check if user exists (and he does not)
+	// 2. We only accept the invitation
+	// But user would register between these two steps
+	// Resulting in a project member not being created for existing user
+	// The raise condition is handled by the PostgreSQL function check_accepted_invitations_for_registered_users()
```

Comment:
Might be a bit overkill, but love the explanation and diligence. 💪 
(also, you probably mean race condition, right?)

Reply:
> (also, you probably mean race condition, right?)

haha, yes 🙈